### ISSUE TITLE
Vibration bugs when two DualSence gamepads connected

### ISSUE THREAD

**Issue by LeXxA3** [Ryochan7/DS4Windows#2604]
DS4Windows 3.1.6, latest HidHide and ViGemBus drivers.
When two dualsense controllers are connected and both are emulated as dualshock 4 (all the settings are set to passthru), in the game - playing a vibration on one controller causes the same vibration on the other, and vice versa. And almost immediately a problem of constant vibration on one of the controllers appears. 
Tested in It Takes Two (Origin).

Passthru light is also works strange, it is black or incorrect color. In most cases it works correct after DS4Windows restart when the gameplay of the game already started.

**Comment by Kanuan** [comment]
Just tested and can confirm.

I've tested with 1xDualSense/1xDS4, and also with 1xDS4/1xDS3(recognized as a DS4).

I tested as the following:

1. Connected 2 controllers
2. Make them use the same or different profiles, doesn't matter as long as they are set for virtual DS4
3. Tested rumble on the [Gamepad Tester Website](https://gamepad-tester.com/)

The rumble command would make a random controller start rumbling, and it wouldn't stop unless clicking on the windows taskbar (which makes the controller stop rumbling because of a completely unrelated issue but anyway...)

**Comment by Yohoki** [comment]
Oof.... my bad... I should have noticed this. I was testing with 2 different controllers on a lot of games, but I never thought to try both at once. All my tests were singleplayer.

**Comment by Ryochan7** [comment]
Just tested this and it looks like the new AwaitRawOutputReport API call only works correctly with one controller. Despite the call being linked to different IDualShock4Controller instances, it seems like it uses shared data so an event on one controller will fire events for the others as well. Might just scrap the idea of virtual DS4 feedback again.

**Comment by Ryochan7** [comment]
Disabled again in commit 3987ed9e4eb9bc87eb6d1cd92bd7cb40a6fc6c49. I tried.

**Comment by nefarius** [comment]
I'm looking into it, folks, no need to be so dramatic 😆

**Comment by nefarius** [comment]
Alright, debug machine prepared, let's investigate!

![image](https://user-images.githubusercontent.com/286631/194884317-aa7f7368-1f9a-4931-bc70-b5ba8803e3d8.png)

**Comment by nefarius** [comment]
Found two issues at play here, one is fixed, the other is a bit more complex, more to come soon.

**Comment by nefarius** [comment]
Good, this is how it is supposed to look like, now onto fixing the SDKs.

![image](https://user-images.githubusercontent.com/286631/194901978-c26c1d19-52f9-4320-9c9f-07c51cea81e9.png)

**Comment by nefarius** [comment]
There we go!

![image](https://user-images.githubusercontent.com/286631/194907483-1bcb9578-47e8-445c-9a32-d55be4231093.png)

@LeXxA3 and whoever can also chime in to test; please follow this:

- Make sure you're on DS4Windows **version 3.1.6**
- Download and overwrite [this DLL](https://ci.appveyor.com/api/buildjobs/4wf4s2hnuo3fb7t3/artifacts/bin%2Fnet6.0%2FNefarius.ViGEm.Client%2FNefarius.ViGEm.Client.dll) in the app directory
- Re-run your tests
- Report back here

Cheers 🥂

**Comment by Yohoki** [comment]
Does this issue only affect mixed pairs of controllers (Dualsense+DS4 / DS3+DS4)? I'm not seeing this bug with 2 or 3 DS4s connected at once, all in DS4 mode. I haven't yet installed the new DLL. I wanted a base check to compare.

**Comment by nefarius** [comment]
> Does this issue only affect mixed pairs of controllers (Dualsense+DS4 / DS3+DS4)? I'm not seeing this bug with 2 or 3 DS4s connected at once, all in DS4 mode. I haven't yet installed the new DLL. I wanted a base check to compare.

I personally tested using a DS4 Rev1 and a DualSense connected. For more combinations, people will need to help and report back.

**Comment by Yohoki** [comment]
> > Does this issue only affect mixed pairs of controllers (Dualsense+DS4 / DS3+DS4)? I'm not seeing this bug with 2 or 3 DS4s connected at once, all in DS4 mode. I haven't yet installed the new DLL. I wanted a base check to compare.
> 
> I personally tested using a DS4 Rev1 and a DualSense connected. For more combinations, people will need to help and report back.

Multiple controller rumble bug -
🟢 Seems to work fine for **DS4 Controllers only**
I just tested with 3 DS4s connected. Each controller seems to vibrate individually. Bug doesn't seem present on the older DLL here. Tested with 2xRev1 and 1xRev2. May only effect mixed pairs of controllers (Dualsense+DS4 / DS3+DS4).

Infinite Rumble -
🟢 Single DS4 (In-Game) - No inf rumble noticed
Rumble seems to work fine with single DS4 controller plugged in while in-game.

🟢 Single DS4 (Rumble Test Buttons) - No inf rumble noticed
Rumble seems to work fine with single DS4 controller plugged in while in DS4Windows' Profile editor.

🟢 Multiple DS4 (Rumble Test Buttons) - No inf Rumble noticed
Rumble also seems to work fine with multiple DS4 controllers plugged in with DS4Windows' Profile Editor.

🔴 Multiple DS4 (Old DLL) (In-Game) - Inf Rumble noticed
Inf rumble noticed while playing 'Detroit: Become Human' with 2 DS4 controllers plugged in while using older DLL. It does not happen immediately, or even on the first rumble event. But did happen within 5 minutes and only stopped when switching tasks.

🟢 Multiple DS4 (Old DLL) (In-Game) - No Inf Rumble noticed
Same setup, newer DLL. Finished a chapter in 'Detroit: Become Human' with no inf rumble.

**Comment by nefarius** [comment]
Soooooo... No issues with the new DLL, am I reading that right?

**Comment by Yohoki** [comment]
Not from me. But I don't have the same setup as OP, Kanuan or yourself. The only issue that seemed to affect me was the inf rumble with multiple controllers, and that looks fixed here.

Some one with a DualSense should test as well, since it looks like you may need different controller to get the bug.

**Comment by Yohoki** [comment]
Could @LeXxA3  and @Kanuan do a quick test in 3.1.6, please? You've both noticed the issue but not said a word on if it has been fixed with the updated ViGEm Driver. Please make sure to test with the same hardware you used to begin with (Ie, a mixed pair of DS4 + Dualsense/DS3)

@nefarius has provided a fix for us. If we can show it's working, then we might be able to get Rumble re-enabled in the next update.

* Make sure you're on DS4Windows v.3.1.6 (**NOT 3.1.7**)
 
* Download and overwrite [this DLL](https://ci.appveyor.com/api/buildjobs/4wf4s2hnuo3fb7t3/artifacts/bin%2Fnet6.0%2FNefarius.ViGEm.Client%2FNefarius.ViGEm.Client.dll) in the DS4Windows directory

@nefarius Did you test only in a debugging tool, or have you tested your mixed controllers in DS4Windows as well? Not that I don't trust your results, just trying to get the info out there. :)

**Comment by andreycout** [comment]
Just tested with a combination of DualSense and DS4 v2. Both via Bluetooth, using the same profile and emulating DS4.
Am on 3.1.6
Tested on https://gamepad-tester.com/
Before swapping the DLL, single controller would vibrate correctly, but after two were connected, testing vibration on one of them would either trigger the other or both and they'd be stuck vibrating. Behavior was kind of random, couldn't verify the pattern.

After swapping the DLL, vibration seems to work fine. Only the selected controller vibrates when I click the vibration test and they didn't get stuck after various tries. So yeah, I think the dll solved the issue.

**Comment by h3nrytan** [comment]
i've tested with two DS4 REV2 controllers via bluetooth emulating DS4 on 3.1.6 with updated DLL, solved the issue of infinite vibration. thanks!

**Comment by nefarius** [comment]
> @nefarius Did you test only in a debugging tool, or have you tested your mixed controllers in DS4Windows as well? Not that I don't trust your results, just trying to get the info out there. :)

All tests have been made with DS4Windows, not some Debug tool.

**Comment by Yohoki** [comment]
@andreycout @h3nrytan: Thanks for the help, guys!

> All tests have been made with DS4Windows, not some Debug tool.
Thanks for the info! It wasn't clear, since the only posts you made had this program and no mention of D4W: https://user-images.githubusercontent.com/286631/194907483-1bcb9578-47e8-445c-9a32-d55be4231093.png
Just wanting to get as much info about the result IN D4W as possible, since the latest update has it disabled.

@Ryochan7, I gathered up a few reviews with the updated ViGEm client. The issues here seem to be fixed with the latest commits to the ViGEm Client SDK, although no releases have been published yet to the SDK or ViGEmBus.

Hopefully that helps. :)

**Comment by Yohoki** [comment]
> > Thanks for the info! Just wanting to get as much info about the result IN D4W as possible, since the latest update has it disabled. It wasn't clear, since the only posts you made had this program and no mention of D4W:
> 
> The process name is literally DS4Windows in the screenshot but I digress 😛
> 
> I'll publish releases tomorrow.

I suppose that's true. lol I'm not familiar with DebugView or how it works. But the clarification helps, nonetheless.

Thanks for the heads up and the quick fix.

**Comment by nefarius** [comment]
[PR created](https://github.com/Ryochan7/DS4Windows/pull/2634)!

And a little tip for the future: if something like this happens again, how about **somebody ping me so that I get informed about the problem?!** 🤣 If it wasn't for @Kanuan I wouldn't have a clue you guys found this issue so yeah, kindly make some noise next time or things won't get resolved properly 😛 

Cheers

**Comment by nefarius** [comment]
### A little Post-Mortem summary for those interested

Got two problems that caused this:

1. I put some code that was supposed to be run on success into the failure handler so it never got executed, oops 😇 
2. Due to limitations of the legacy design of ViGEmBus and a component I use ([`Dmf_NotifyUserWithRequestMultiple`](https://github.com/Microsoft/DMF/blob/master/Dmf/Modules.Library/Dmf_NotifyUserWithRequestMultiple.md)) separating the output reports has to be done by taking the emulated device's serial index into account. I fixed that by spawning a thread which takes care of de-queueing all pending requests and the properly dispatch them to the device-specific wait functions where the packets got reported back to the correct device in the right order again.

So yeah, the older some code gets the harder it becomes to extend without breaking everything 😅

**Comment by Yohoki** [comment]
I don't know the codes well enough to say which is an issue from D4W and which is an issue from ViGEm, so it's hard to know where to put an issue here.

Actually, I've been encountering another issue, that claims is a ViGEm Error in the logs. I put a discussion up about it, but wasn't sure if I should tag you or Ryo. I'll give a tag in it and let you take a quick read, i guess.

**Comment by nefarius** [comment]
> I don't know the codes well enough to say which is an issue from D4W and which is an issue from ViGEm, so it's hard to know where to put an issue here.
> 
> Actually, I've been encountering another issue, that claims is a ViGEm Error in the logs. I put a discussion (#2564) up about it, but wasn't sure if I should tag you or Ryo. I'll give a tag in it and let you take a quick read, i guess.

The core maintainers *should* 😉

**Comment by LeXxA3** [comment]
Hello @nefarius , @Yohoki ,

Sorry, missed your reqeust. Yes, vibration with new DLL is fine. Tested in It Takes Two with 2 DualSense in DS4 emulation.
But I found new small bug with new DLL:
When you press stop button - DS4windows stops, but I can't start it, button remains "Stop" and it is unclickable:
![stopped](https://user-images.githubusercontent.com/25710897/196023356-fc56b2db-b22b-427f-bb92-eefb5662600b.PNG)

Not reproducible when I use old DLL from 3.1.6.

**Comment by nefarius** [comment]
> Hello @nefarius , @Yohoki ,
> 
> Sorry, missed your reqeust. Yes, vibration with new DLL is fine. Tested in It Takes Two with 2 DualSense in DS4 emulation. But I found new small bug with new DLL: When you press stop button - DS4windows stops, but I can't start it, button remains "Stop" and it is unclickable: ![stopped](https://user-images.githubusercontent.com/25710897/196023438-8b1ffee0-5ad3-4f6b-8da2-29430ff66fcc.PNG)
> 
> Not reproducible when I use old DLL from 3.1.6.

Thanks, we also discovered and worked on that in parallel, I'll update the PR accordingly later today. In the meantime you can try this DLL https://ci.appveyor.com/api/buildjobs/rc6fu7v1lx7b8qub/artifacts/bin%2Fnet6.0%2FNefarius.ViGEm.Client%2FNefarius.ViGEm.Client.dll

**Comment by LeXxA3** [comment]
With new DLL vibration is fine and "start/stop" button is also fine, thank you so much for awesome work : )

**Comment by nefarius** [comment]
Nice, thanks for confirming, will update the PR later and then all should be well 😁

**Comment by andreycout** [comment]
> Hello @nefarius , @Yohoki ,
> 
> Sorry, missed your reqeust. Yes, vibration with new DLL is fine. Tested in It Takes Two with 2 DualSense in DS4 emulation. But I found new small bug with new DLL: When you press stop button - DS4windows stops, but I can't start it, button remains "Stop" and it is unclickable: ![stopped](https://user-images.githubusercontent.com/25710897/196023438-8b1ffee0-5ad3-4f6b-8da2-29430ff66fcc.PNG)
> 
> Not reproducible when I use old DLL from 3.1.6.

Oh, so that's what was causing it. I've been working around it by right clicking the tray icon and clicking Stop (which funnily enough actually starts it when it's stopped and the button is unclickable. It has been stopping itself after resume from pc sleep for some reason)

Will test this new dll, thanks @nefarius

**Comment by nefarius** [comment]
The PR has been updated and includes this fix as well. To anybody who likes to continue testing while we wait for review; [here is the latest DLL](https://ci.appveyor.com/api/buildjobs/l9makps170jnd7o6/artifacts/bin%2Fnet6.0%2FNefarius.ViGEm.Client%2FNefarius.ViGEm.Client.dll) you can drop in like mentioned above.

**Comment by LeXxA3** [comment]
> The PR has been updated and includes this fix as well. To anybody who likes to continue testing while we wait for review; [here is the latest DLL](https://ci.appveyor.com/api/buildjobs/l9makps170jnd7o6/artifacts/bin%2Fnet6.0%2FNefarius.ViGEm.Client%2FNefarius.ViGEm.Client.dll) you can drop in like mentioned above.

This is strange. Is this for 3.1.6 ?
When I run DS4Windows 3.1.6 with this DLL, program just closing without errors.

**Comment by nefarius** [comment]
> > The PR has been updated and includes this fix as well. To anybody who likes to continue testing while we wait for review; [here is the latest DLL](https://ci.appveyor.com/api/buildjobs/l9makps170jnd7o6/artifacts/bin%2Fnet6.0%2FNefarius.ViGEm.Client%2FNefarius.ViGEm.Client.dll) you can drop in like mentioned above.
> 
> This is strange. Is this for 3.1.6 ? When I run DS4Windows 3.1.6 with this DLL, program just closing without errors.

I replaced the link with a more recent one, can you try again please? Also, do you see any recent errors in Event Viewer / Application log?

**Comment by nefarius** [comment]
Whoever reads this: **we need more testers and feedback**, where are the people at?! Fulfill your community duties if you want to see this project improve 😉

**Comment by Masamune3210** [comment]
Only have one functional DS4 or I would

**Comment by Ryusennin** [comment]
I also had a silent crash with the previous client DLL build, but the latest update (1.21.230) runs fine with DS4W 3.1.6.

I only have one DS4 v2 at my disposal, and the only game that produced infinite rumble was Spider-Man (when interacting with doors, devices, computers, etc). I had to pause the game to reset the rumble. Nefarius' library fixed that issue completely.

**Comment by Yohoki** [comment]
> Only have one functional DS4 or I would

All feedback is important. Even a single controller's behavior. It's only because of people trying different things that the bug was discovered to begin with. New people find things that we never thought of when coding.

**Comment by Masamune3210** [comment]
Well, I'm about done with the first Dragon Age, I'll use the new dll when I start 2 and report anything that might turn up. Or is that too old of a game to be helpful information

**Comment by Masamune3210** [comment]
I can also try with my DualSense, I recently figured out its issue and got it working again so I now have one of each

**Comment by nefarius** [comment]
> Or is that too old of a game to be helpful information

The game age has nothing to do with testing **correct functionality** which is what we're after. Test whatever scenario you like as long as you post exactly what you did with what version and which devices.

**Comment by andreycout** [comment]
> The PR has been updated and includes this fix as well. To anybody who likes to continue testing while we wait for review; [here is the latest DLL](https://ci.appveyor.com/api/buildjobs/abblyteff3trc1vq/artifacts/bin%2Fnet6.0%2FNefarius.ViGEm.Client%2FNefarius.ViGEm.Client.dll) you can drop in like mentioned above.

Just tested this DLL with 3.1.6. I was also getting the insta-crash reported by @LeXxA3, but figured out it was caused by Windows permission issue. DLL files downloaded directly from the internet will be "locked", you have to right click them, select properties and tick the "unlock file" checkbox (not sure about the exact name as my Windows is not in English). For whatever reason, even though I'm admin, even after confirming the UAC prompt, Windows would fail to overwrite the permission (something about declaration failure).

What I did to fix it was: redownload the DLL, but now I chose to "save file as" and renamed the extension to .txt before clicking download. After the download was completed, I renamed from .txt do .dll and THEN went to properties and chose to "unlock" the file. It worked this time.

Tested again with one DualSense and one DS4 v2, using gamepad tester website. No vibration bug after moderate amount of testing, both emulating DS4, and switching back and forth between DS4 emulation and Xbox 360 controller emulation didn't glitch/crash or cause the bug to occur.

TLDR: DLL posted by @nefarius works fine on my end

**Comment by rafa21k1** [comment]
En mi experiencia use la DLL en 3.1.6 y se me crasheaba el sistema en win11, pasé a win 10 y lo mismo, dejé la anterior DLL e iba todo bien

**Comment by nefarius** [comment]
> > The PR has been updated and includes this fix as well. To anybody who likes to continue testing while we wait for review; [here is the latest DLL](https://ci.appveyor.com/api/buildjobs/abblyteff3trc1vq/artifacts/bin%2Fnet6.0%2FNefarius.ViGEm.Client%2FNefarius.ViGEm.Client.dll) you can drop in like mentioned above.
> 
> Just tested this DLL with 3.1.6. I was also getting the insta-crash reported by @LeXxA3, but figured out it was caused by Windows permission issue. DLL files downloaded directly from the internet will be "locked", you have to right click them, select properties and tick the "unlock file" checkbox (not sure about the exact name as my Windows is not in English). For whatever reason, even though I'm admin, even after confirming the UAC prompt, Windows would fail to overwrite the permission (something about declaration failure).
> 
> What I did to fix it was: redownload the DLL, but now I chose to "save file as" and renamed the extension to .txt before clicking download. After the download was completed, I renamed from .txt do .dll and THEN went to properties and chose to "unlock" the file. It worked this time.
> 
> Tested again with one DualSense and one DS4 v2, using gamepad tester website. No vibration bug after moderate amount of testing, both emulating DS4, and switching back and forth between DS4 emulation and Xbox 360 controller emulation didn't glitch/crash or cause the bug to occur.
> 
> TLDR: DLL posted by @nefarius works fine on my end

Right, it's called the Zone Identifier, a useless piece of "security" that browsers set on executables. I always forget about it, have it disabled on all my systems.

**Comment by Masamune3210** [comment]
Unfortunately pretty much everybody other than Android and Linux use the idea in one way or another, not that it particularly helps anything other than getting in the way. All it does on Windows is cue the extra dialog when doing certain operations that tell the user that the file shouldn't be trusted, cause some stronger checks on loading and moving files, and cause AV to be triggered to check the file
